stan/math/gpu/cholesky_decompose.hpp

```diff
@@ -0,0 +1,130 @@
+#ifndef STAN_MATH_GPU_CHOLESKY_DECOMPOSE_HPP
+#define STAN_MATH_GPU_CHOLESKY_DECOMPOSE_HPP
+#ifdef STAN_OPENCL
+#include <stan/math/gpu/matrix_gpu.hpp>
+#include <stan/math/gpu/kernels/cholesky_decompose.hpp>
+#include <stan/math/gpu/multiply.hpp>
+#include <stan/math/gpu/multiply_transpose.hpp>
+#include <stan/math/gpu/lower_tri_inverse.hpp>
+#include <stan/math/gpu/transpose.hpp>
+#include <stan/math/gpu/subtract.hpp>
+#include <stan/math/gpu/err/check_diagonal_zeros.hpp>
+#include <stan/math/gpu/err/check_nan.hpp>
+#include <CL/cl.hpp>
+#include <algorithm>
+#include <cmath>
+
+namespace stan {
+namespace math {
+inline matrix_gpu cholesky_decompose(matrix_gpu& A, const int block = 100,
```

Comment:
I don't think these default values make sense - we want to recursively divide up the input into smaller and smaller chunks. This is related to the other comments e.g. https://github.com/stan-dev/math/pull/1059/files#r236081592. I don't think I totally get the way that while loop on line 81 in this file is supposed to interact with the block size and would suggest an entirely recursive algorithm without that while loop. Also need to see tests for different weird values of these parameters.

Reply:
If you want to support an API that doesn't require a block size passed in yet allows it to be configured, I would suggest using some kind of sentinel value (like block<1 or something) that would indicate that the function is responsible for setting an appropriate block size based on A. But I wonder if we actually ever want to pass in a block size at all given that it seems the most efficient one can be immediately determined from the size of A?

Reply:
I was working on this refactor but then had other stuff, then I took some time of around new years. Its a bit tricky to make a 

I agree with your comment on default values. 

We can determine an efficient value based on experience yes. Its hard to say if its the most efficient, but its definitely very close at least on recent hardware.

Reply:
@seantalts I've tried figuring out how to get rid of this while loop but I'm not really seeing how to do it. Would you be able to take a stab at it or sit down with me to work this out? We can also do a google hangout so @rok-cesnovar can join as well.

Unless Rok has figured it out then we can wait a bit

Reply:
Excellent much appreciated!